### DISCUSSION TITLE
Confusion about force calculation (Issues related to development)

### DISCUSSION THREAD

**Discussion by lhycms** [deepmodeling/deepmd-kit#3014] [Q&A] [answered]
I try to figure out how the nn model calculate forces on atoms. I usually establish neighbor list with ASE, and calculate forces acoording to below:
![截屏2023-11-28 下午5 43 49](https://github.com/deepmodeling/deepmd-kit/assets/41718895/fc7497cc-2cba-4778-82e3-d1713b5545fd)

But lammps estiblish neighbor list in different way, and we must take forces on ghost atoms in account when use lammps-mpi.
[
![截屏2023-11-28 下午5 43 56](https://github.com/deepmodeling/deepmd-kit/assets/41718895/214c850b-9d0c-4306-8b58-8d7a0f57ab0e)
](url)

I don't know how to calculate the derivative of ghost atom energy with respect to relative coordinate. 

Could anyone can help me solve it? Thanks.

**Comment by njzjz** [comment]
> I don't know how to calculate the derivative of ghost atom energy with respect to relative coordinate.

In the DeePMD-kit, ghost atoms do not contribute atomic energy, so this part is zero.

**Reply by lhycms** [reply]
Thanks for the help😄

I still have three questions:

1. If ghost atoms do not contribute atomic energy, the approach taken in figure#1 is completely incorrect, i shouldn't map ghost atom index to its primitive index and add corresponding partial derivative term, is that so?

2. In case shown as above, beceuse atom#2 is outside the cutoff of atom#1, and atom#1 is outside the cutoff of atom#2. So the correct calculation formula for force is: $F_1=\frac{\partial E_1}{\partial R_{13}} + 0$, $F_2=\frac{\partial E_2}{\partial R_{24}}+0$, Is it?

3. I tried to derive the formula for atomic forces, which is different from `Computer Physics Communications, 2018, 228: 178-184`. If you have time, i really hope you can help me check this process as below:
<img width="1027" alt="截屏2023-11-30 上午1 00 04" src="https://github.com/deepmodeling/deepmd-kit/assets/41718895/e59e2d67-fb5d-4e19-b826-8e6f0bd450d4">

**Reply by njzjz** [reply]
I am not sure whether you want to discuss $F_i$ before communication or after communication. Maybe you should clarify it. It seems $F_i$ in Figure 1 is the term **after** communication (which is correct since there is no communication in ASE), while $F_i$ in Figure 2 is the term **before** communication. They are expected not to be the same.

**Reply by lhycms** [reply]
Thanks again for you reply!

I'm sorry that not clarify the issue about communication. Now i got what you said:
1. When using single core, there is no concept about `communication`, the portions surrounded by the red dashed lines will be added onto the forces directly.
2. When uising multiple cores, the portions surrounded by the red dashed lines will be added onto the forces during the reverse communication process.

The dimension of matrix $\frac{E_{tot}}{R_{ij}}$ should be [nframes, nlocal, nneigh, 3]. In above case(figure#2), `nframes=1, nlocal=2, nneigh=1, nxyz=3`, so dimension of $\frac{E_{tot}}{R_{ij}}$ is [1, 2, 1, 3], we can easliy get $\frac{E_1}{R_{13}}$ using index [0, 0, 0, :] and $\frac{E_2}{R_{24}}$ using index [0, 1, 0, :]. 
But the corresponding index of $\frac{\partial E_3}{\partial R_{31}}$ should be [0, 2, 0, :], which is beyond boundary in second dimension. Does deepmd-kit use some map that can map $\frac{E_{3}}{R_{31}}$ to $\frac{E_{2}}{R_{24}}$ whose corresponding index is [0, 1, 0, :]? If not so, we seem can't get $\frac{\partial E_3}{\partial R_{31}}$. This problem still confuse me.

**Reply by njzjz** [reply, marked as answer]
As I mentioned above, $E_3$ is zero.

**Reply by lhycms** [reply]
Thank you for your prompt reply. I understand this issue.

Wish you all the best in your work! 😄
